pkg/cache/sql/informer/informer.go

```diff
@@ -0,0 +1,82 @@
+/*
+package sql provides an Informer and Indexer that uses SQLite as a store, instead of an in-memory store like a map.
+*/
+
+package informer
+
+import (
+	"context"
+	"time"
+
+	"github.com/rancher/lasso/pkg/cache/sql/partition"
+	sqlStore "github.com/rancher/lasso/pkg/cache/sql/store"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+)
+
+// Informer is a SQLite-backed cache.SharedIndexInformer that can execute queries on listprocessor structs
+type Informer struct {
+	cache.SharedIndexInformer
+	ByOptionsLister
+}
+
+type ByOptionsLister interface {
+	ListByOptions(ctx context.Context, lo ListOptions, partitions []partition.Partition, namespace string) (*unstructured.UnstructuredList, string, error)
+}
+
+// NewInformer returns a new SQLite-backed Informer for the type specified by schema in unstructured.Unstructured form
+// using the specified client
+func NewInformer(client dynamic.ResourceInterface, fields [][]string, gvk schema.GroupVersionKind, db sqlStore.DBClient, shouldEncrypt bool, namespaced bool) (*Informer, error) {
+	listWatcher := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			a, err := client.List(context.TODO(), options)
+			return a, err
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return client.Watch(context.TODO(), options)
+		},
+	}
+
+	example := &unstructured.Unstructured{}
+	example.SetGroupVersionKind(gvk)
+
+	// avoids the informer to periodically resync (re-list) its resources
+	// currently it is a work hypothesis that, when interacting with the UI, this should not be needed
+	resyncPeriod := time.Duration(0)
+
+	sii := cache.NewSharedIndexInformer(listWatcher, example, resyncPeriod, cache.Indexers{})
+
+	name := informerNameFromGVK(gvk)
+
+	s, err := sqlStore.NewStore(example, cache.DeletionHandlingMetaNamespaceKeyFunc, db, shouldEncrypt, name)
+	if err != nil {
+		return nil, err
+	}
+	loi, err := NewListOptionIndexer(fields, s, namespaced)
+	if err != nil {
+		return nil, err
+	}
+
+	// HACK: replace the default informer's indexer with the SQL based one
+	UnsafeSet(sii, "indexer", loi)
+
+	return &Informer{
```

Comment:
From what I can tell here, what we are doing here is the following:
- We are constructing a `sharedIndexInformer`. 
- Our `Informer` struct is implementing the `SharedIndexInformer` interface.
- We are overriding the internal `indexer` field of the struct to be our new indexer, which also uses a SQL store. 
- We are implementing this `ListByOptions` function, which allow us to be more creative in how the SQL store is accessed.

I've got a couple of overall questions on this:
- SharedIndexInformer is an interface. Rather than trying to override an unexported field on the concrete type, could we instead implement our own version of this interface?
  - Internally, this could construct a `SharedInformer` and use `AddEventHandler` to allow us to index these types without implementing the run loop.
    - This does have a memory cost since the original informer would still store objects.
    - We could vendor the run loop at the moment to avoid this.
- The listOptions/Partition logic seems like something that belongs to a consuming library. It's using the indexer for enhanced querying, but this isn't logic that helps to run a controller, and it doesn't conform to the indexer interface. Can you elaborate on how this fits within Lasso? 
  - To go a bit further, it looks like the SharedInformer would allow us to work with a store that doesn't have indexers. If we are going to use something other than an indexer to query values, it seems like it would make sense to use that interface rather than the SharedIndexerInformer one. Though I might be missing what we get out of the indexer in this model.

Reply:
Can you clarify what you mean when you say we could construct a SharedInformer internally? Are you proposing we still use the sharedIndexInformer type but wrap it and expose a run that just uses event handler?

We provide informer backed caches in lasso. The store that we are extending now provides new functionality which forms queries based on the the options struct. In that sense I think it's the best place to provide it. It also supports the concept of partitions as priority zero filters for the store. I'm not entirely sure how any of this does not fit in lasso. Sure there are other ways we can move the structs among our 3 to 4 repos but I'm not sure it gets any more logical.

Reply:
About pairing our SQL-based `Store` to a `SharedInformer` instead of `SharedIndexInformer` to trim all `Indexer` related fat we do not actually use: it'd be a great idea!

There's a catch though. The problem is we'd have to replicate some logic from client-go's implementation, because the provided `SharedInformer` implementation is simply using a `SharedIndexInformer` instance implementation with an empty `Indexer` list under the hood 😢:

https://github.com/rancher/client-go/blob/b4d94f01600c2edfda43d82052bb0a14d0bdb120/tools/cache/shared_informer.go#L162-L163

Reimplementing the `SharedInformer` interface, then, would mean IIUC replicating ~600 lines from shared_informer.go plus tests and dependencies (from the definition of `sharedIndexInformer` on).

I do not think we want to go down that particular path (note: implementing something other than `SharedInformer` is different).

Reply:
> Reimplementing the SharedInformer interface, then, would mean IIUC replicating ~600 lines from shared_informer.go plus tests and dependencies (from the definition of sharedIndexInformer on).

Agreed. The way I see it, if we want a `cache.SharedIndexInformer` then we're better off re-using the one from client-go instead of re-implmenting those 600+ lines (via unsafe hack/client-go fork patch). I think the `Store`+`Controller` approach would be useful in the case that we don't want the `cache.SharedIndexInformer`.